pkg/buildqueue/buildqueue.go

```diff
@@ -0,0 +1,568 @@
+type buildStatus struct {
+	builder Builder
+
+	// curentBuild is the building information of any running build
+	currentBuild *build
+
+	// nextBuild is the building information of the next build for this
+	// UUID
+	nextBuild *build
+}
+
+func (q *BuildQueue) enqueueBuild(b Builder) (BuildNotification, bool) {
+	build, enqueued := q.serializeBuild(b)
+	notify := build.createAndAddNotificationChannel()
+
+	metrics.BuildQueueEntries.With(map[string]string{
+		metrics.LabelBuildQueueName:  q.name,
+		metrics.LabelBuildQueueState: metrics.BuildQueueWaiting,
+	}).Inc()
+
+	b.BuildQueued()
+
+	return notify, enqueued
+}
+
+func (q *BuildQueue) serializeBuild(b Builder) (*build, bool) {
+	uuid := b.GetUUID()
+	build := newBuild(b, uuid)
+
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+
+	if buildStatus, ok := q.buildStatus[uuid]; ok {
+		// If the builder is currently being built, prepare the next
+		// build and store it in the nextBuild field. When the current
+		// build finishes, the next build will automatically be added
+		// to the workerBuildQueue
+		if buildStatus.currentBuild != nil && buildStatus.nextBuild == nil {
+			buildStatus.nextBuild = build
+			return build, false
+		}
+
+		// The builder is already in the queue but not being built, the
+		// build request will be fulfilled when the queued build is
+		// executed. The caller can be notified to skip the build.
+		return buildStatus.nextBuild, false
+	}
+
+	q.buildStatus[uuid] = &buildStatus{
+		builder:   b,
+		nextBuild: build,
+	}
+
+	return build, true
+}
+
+// Enqueue schedules Builder for building. A channel is returned to provide the
+// ability to wait for the build to complete and check for success or failure.
+//
+// A regular build fullfils the following build guarantees:
+// * The same Buildable, i.e. the same UUID, will not be built in parallel on
+//   multiple workers.
+// * The order in which regular builds are enqueued is maintained in the build
+//   order across different Buildables. If the same Buildable is enqueued
+//   multiple times, the buils are automatically folded together until the
+//   build is executed. This means in practise that a specific Buildable can only
+//   be in the build queue once.
+// * In the presence of preemptive, exclusive builds, regular builds get a fair
+//   chance to run but preemptive builds always preempt any queued, not yet
+//   running regular build.
+func (q *BuildQueue) Enqueue(b Builder) BuildNotification {
+	notify, enqueue := q.enqueueBuild(b)
+	if enqueue {
+		// This should be non-blocking unless there is contention beyond
+		// queueBuildSize in which case this will block until a slot in the
+		// queue becomes available.
+		q.workerBuildQueue <- b
+	}
+
+	return notify
+}
+
+// Remove removes the builder from the queue.
+func (q *BuildQueue) Remove(b Buildable) {
+	uuid := b.GetUUID()
+	q.mutex.Lock()
+	if status, ok := q.buildStatus[uuid]; ok {
+		delete(q.buildStatus, uuid)
```

Comment:
The build might be in progress. Shouldn't we keep the builder in the map in that case, until the build completes? Or maybe it would even be simpler to keep the builder in the map if `currentBuild` is not `nil`, since it would be deleted anyway after the build completes?

I think it works now, since `needToBuildAgain()` handles the deletion from the map well, but it might be surprising. In the normal case, a build may be deleted from the map only after a build has completed and there is no need to rebuild.

Reply:
Yes, I agree that is more correct. The build status is guaranteed to be cleared up after the build completes.

Reply:
I have changed the code to keep the buildStatus if a build is ongoing and defer the deletion to `needToBuildAgain()`.